Added check for unsupported dispatch key in codegen

francescocastelli opened the pull request:
Summary: Added a check for the dispatch keys present in native_function.yaml, they must be part of the fixed set of dispatch keys. If not, signal an error. I also removed two dispatch keys from the function schema copy_ , because they are not supported (SparseHIP, SpareXPU).

Test: this function schema (for example) in native_function.yaml 
```
- func: native_norm(Tensor self, Scalar p=2) -> Tensor
  dispatch:
    SparseCPU, SparseCUDA, SparseHIP: norm_sparse
```
now generates this error during codegen:  `AssertionError: SparseHIP is not a supported dispatch key.` 

Fixes #66190

ezyang commented on tools/codegen/gen.py:
```diff
@@ -1069,6 +1069,11 @@ def gen_source_files(
+    # Only the set of keys defined in dispatch_keys should be supported,
+    # raise an error for unsupported keys
+    for k in backend_indices.keys():
+        assert k in dispatch_keys, f"{k} is not a supported dispatch key."
```
This doesn't seem quite right. For one, there is a `--backend-whitelist` which optionally controls what is in `dispatch_keys`, but we ought not fail if `native_functions.yaml` has a backend that wasn't in the whitelist. I'd probably put the check here in model.py and maintain a non-wobbling list of all valid dispatch keys for codegen, e.g., factor out this chunk:

```
    dispatch_keys = [
        DispatchKey.CPU,
        DispatchKey.SparseCPU,
        DispatchKey.SparseCsrCPU,
        DispatchKey.MkldnnCPU,
        DispatchKey.CUDA,
        DispatchKey.SparseCUDA,
        DispatchKey.SparseCsrCUDA,
        DispatchKey.QuantizedCPU,
        DispatchKey.QuantizedCUDA,
        DispatchKey.CompositeImplicitAutograd,
        DispatchKey.CompositeExplicitAutograd,
        # Meta is a magic key: it is automatically generated for structured
        # kernels
        DispatchKey.Meta,
        DispatchKey.ZeroTensor,
    ]
```

ezyang replied:
yeah, this is why your android builds are failing

francescocastelli commented:
@ezyang Hi! I can probably take a look at this next week if this is ok for you

francescocastelli committed:
fixed supported dispatch keys assert

francescocastelli committed:
fixed identation

francescocastelli committed:
fixed identation

francescocastelli commented:
@ezyang I moved the check in model.py. I didn't know where to put the list of supported keys, so I moved it also in model.py, let me know where the best place is!

ezyang reviewed (approved):
Thanks!
